crates/telio-task/src/task/runtime.rs

```diff
@@ -51,127 +53,64 @@ pub trait Runtime: Sized {
-#[async_trait]
-pub trait RuntimeExt: Runtime {
-    /// Guard multiple async operations form being interrupted by exec
-    fn guard<'a, F>(block: F) -> WaitResponse<'a, Self::Err>
-    where
-        F: Future<Output = Result<(), Self::Err>> + Send + 'a,
-    {
-        WaitResponse(block.boxed())
-    }
-
-    /// Continue to next loop iteration
-    fn next() -> WaitResponse<'static, Self::Err> {
-        WaitResponse(ready(Ok(())).boxed())
-    }
-
-    /// Error out with [Runtime::Err]
-    fn error(e: Self::Err) -> WaitResponse<'static, Self::Err> {
-        WaitResponse(ready(Err(e)).boxed())
-    }
-
-    /// Sleep forever in loop, loop is stopped, but not deallocated
-    /// Task loop can be re-triggered using exec
-    async fn sleep_forever() -> WaitResponse<'static, Self::Err> {
-        pending().await
-    }
-}
-
-impl<T> RuntimeExt for T where T: Runtime {}
-
-/// A general runtime for components
-///
+/// A general runtime for compoments.
 /// This task should be used in components requiring long running actions.
 ///
 /// It takes care of:
 ///   * Graceful stop. (On drop or with stop)
 ///   * Wait entry to wait for outside triggers. (Like [Rx], sockets, timers etc)
 ///   * Ability to execute mutation on state without requiring mutex'es
 pub struct Task<S: Runtime> {
-    stop: Arc<Notify>,
-    execute: Tx<Update<S, S::Err>>,
-    join: Option<JoinHandle<Result<(), S::Err>>>,
-}
-
-/// Task was stopped during execution
-#[derive(Debug, thiserror::Error)]
-#[error("Failed to execute.")]
-pub struct ExecError;
-
-/// Task stop status.
-#[derive(Debug)]
-pub enum StopResult<E> {
-    /// Task was stopped successfully
-    Ok,
-    /// Task stopped due to internal error
-    Err(E),
-    /// Task panic'ed
-    Panic(Box<dyn Any + Send + 'static>),
-}
-
-/// Response from [Runtime::wait].
-///
-/// It's recommended to use [RuntimeExt] methods.
-pub struct WaitResponse<'a, E>(BoxFuture<'a, Result<(), E>>);
-
-/// Future closure
-pub trait Action<V, R = ()>
-where
-    for<'a> Self: FnOnce(&'a mut V) -> BoxFuture<'a, R> + Send + 'static,
-    R: Send + 'static,
-{
-}
-
-impl<T, V, R> Action<V, R> for T
-where
-    for<'a> T: FnOnce(&'a mut V) -> BoxFuture<'a, R> + Send + 'static,
-    R: Send + 'static,
-{
+    pub(crate) stop: Tx<StopKind>,
+    pub(crate) execute: Tx<Update<S, S::Err>>,
+    pub(crate) join: Option<JoinHandle<Result<StopKind, S::Err>>>,
 }
 
-/// Boxed future closure
-pub type BoxAction<V, R = ()> = Box<dyn Action<V, R>>;
-
-type AnySend = Box<dyn Any + Send + 'static>;
-
-type Resp<T> = oneshot::Sender<T>;
-
-type Update<S, E> = (BoxAction<S, Result<AnySend, E>>, Resp<AnySend>);
-
 impl<S> Task<S>
 where
     S: Runtime + Send + 'static,
 {
     /// Start a new task for state
     pub fn start(mut state: S) -> Self {
-        let stop = Arc::new(Notify::new());
+        let Chan {
+            tx: stop,
+            rx: mut stopped,
+        } = Chan::<StopKind>::new(1);
```

Comment:
maybe a `tokio::sync::oneshot` would be better here, since I expect that only one message should be sent over this channel?